blueprints/flexible-topologies.md

```diff
@@ -0,0 +1,395 @@
+
+### Traffic Ops Impact
+
+Traffic Ops will provide the ability to create Topologies, composed of cachegroups and parent relationships, which will be assignable to one or more Delivery Services.
+
+#### REST API Impact
+
+The following is the JSON representation of a `Topology` object:
+
+```JSON
+{
+    "name": "foo",
+    "description": "a foo topology",
+    "nodes": [
+        {
+            "cachegroup": "child-cachegroup",
+            "parents": [1, 2]
+        },
+        {
+            "cachegroup": "parent-cachegroup",
+            "parents": []
+        },
+        {
+            "cachegroup": "secondary-parent-cachegroup",
+            "parents": []
+        }
+    ]
+}
+```
+
+The following table describes the top-level `Topology` object:
+
+| field       | type                        | optionality | description                                                         |
+| ----------- | --------------------------- | ----------- | ------------------------------------------------------------------- |
+| name        | string                      | required    | a unique name for identifying this Topology                         |
+| description | string                      | required    | the description of this Topology                                    |
+| nodes       | array of `node` sub-objects | required    | the set of `nodes` in this topology, similar to an *adjacency list* |
+
+The following table describes the `node` sub-object:
+
+| field      | type              | optionality | description                                                                                                                                                                                                 |
+| ---------- | ----------------- | ----------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| cachegroup | string            | required    | the `short_name` of a cachegroup this node maps to in the Topology                                                                                                                                          |
+| parents    | array of integers | required    | zero-based indexes to other nodes in the Topology's `nodes` array, where the 1st element is for the *primary* parent relationship and the 2nd element is for the *secondary* parent relationship, and so on |
+
+API constraints:
+- a Topology must have at least 1 `node`; otherwise, it is useless
+- there cannot be multiple `nodes` for the same cachegroup in a Topology
+- `parents` must have 0, 1 or 2 elements, cannot contain duplicates, cannot contain the index of its own `node`, and cannot contain the index of `nodes` whose cachegroup is of type `EDGE_LOC`
```

Comment:
could parents not look like this:

`"parents": [1, 1]"`

which is essentially is the same as "no 2ndary parent" or i suppose its just easier to prevent duplicates.

Reply:
I'm definitely in favor of restricting duplicates.